### PR TITLE
[Model] Support VLMs with transformers backend

### PR DESCRIPTION
**Pull request by zucchini-nlp**
This PR adds support for multimodal models in Transformers backend. As a start I tested with vanilla LLaVA using demo scripts from the documentation. The generated outputs matched with VLLM outputs. 

For this branch to work, we first need a few changes from `transformers` starting from https://github.com/huggingface/transformers/pull/36367. Currently I want to ask for feedback, if this aligns with how VLLM sees things

cc @Isotr0py @ArthurZucker

### CODE REVIEW

File: vllm/model_executor/models/transformers.py [hunk 1]
```diff
@@ -132,12 +307,13 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = "") -> None:
...
-        self.vocab_size = config.vocab_size
-        self.unpadded_vocab_size = config.vocab_size
+        self.text_config = config.get_text_config()
+        self.vocab_size = self.text_config.vocab_size
+        self.unpadded_vocab_size = self.text_config.vocab_size
 
         self.model: PreTrainedModel = AutoModel.from_config(
             self.config,
-            attn_implementation="vllm",
+            attn_implementation={"text_config": "vllm", "vision_config": "eager"},
```
Comment (zucchini-nlp) [inline vllm/model_executor/models/transformers.py:316]: Current way doesn't work with LLMs. Setting `self.config.get_text_config().attn_implementation="vllm"` will be much more generic, but that needs a fix on `transformers` first

File: vllm/model_executor/models/transformers.py [hunk 2]
```diff
@@ -119,10 +124,180 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+            mm_kwargs=mm_kwargs,
+            mm_hashes=None,
+            mm_placeholders=mm_placeholders,
+        )
+
+
+@MULTIMODAL_REGISTRY.register_processor(MultiModalProcessor,
+                                        info=MultiModalProcessingInfo,
+                                        dummy_inputs=MultiModalDummyInputsBuilder)
+class TransformersModel(nn.Module, SupportsQuant, SupportsMultiModal):
```
Comment (jeejeelee) [inline vllm/model_executor/models/transformers.py:298]: Using the same class to support both text-only models and multimodal models makes it difficult to maintain V1 compatibility, see: https://github.com/vllm-project/vllm/pull/13157

**Comment by DarkLight1337** [comment]
Thanks for working on this! The main difficulty of supporting VLMs is not the model implementation itself, but rather the preprocessing code. In particular, vLLM V1 requires precise tracking of the placeholder tokens. I see how generalizing `return_token_type_ids` to the multimodal context can help. We still have a couple issues to tackle though:
- vLLM doesn't support non-consecutive multimodal placeholder feature tokens, which occurs when image tokens are split into rows and columns by other padding tokens. We currently work around this in vLLM by considering those padding tokens as feature tokens as well, as demonstrated in our [Fuyu example](https://docs.vllm.ai/en/latest/contributing/model/multimodal.html). To enable this workaround, we also require information about the padding tokens when `return_mm_token_type_ids=True`.
- vLLM supports text+multimodal and token+multimodal inputs, but current HF processors only support the former. Is it feasible to also support token+multimodal inputs in HF Transformers?

cc @ywang96

**Comment by zucchini-nlp** [comment]
@DarkLight1337 thanks for review! Yeah, checking on more involved models is a good idea to verify all edge cases are covered, will do so. A few clarifications before that:
- For LLaVA-OneVision do we need to support inputs with video and image within one batch? If yes, that would complicate things a bit probably
- When doing `token+multimodal inputs` for the processor, do we expect the tokens to be already expanded or anyhow processed for multimodality? Or we assume that tokens are simply text+tokenizer output? I am not sure if it is a good idea to support it as part of `call`, but we can add a private method for VLLM to use. Also it would give us freedom to change API without caring for BC in the future, if we decide to handle `tokens` differently

**Comment by DarkLight1337** [comment]
> For LLaVA-OneVision do we need to support inputs with video and image within one batch? If yes, that would complicate things a bit probably

Yes, we current support mixed-modality (non-interleaved) inputs and plan to eventually support interleaved-modality inputs as well.

> When doing token+multimodal inputs for the processor, do we expect the tokens to be already expanded or anyhow processed for multimodality? Or we assume that tokens are simply text+tokenizer output? I am not sure if it is a good idea to support it as part of call, but we can add a private method for VLLM to use. Also it would give us freedom to change API without caring for BC in the future, if we decide to handle tokens differently

We assume that the tokens have only gone through the tokenizer. So, prompt replacement still has to be performed on the tokens.

**Comment by ywang96** [comment]
Thanks for the PR @zucchini-nlp! I'm a bit occupied at the moment but will take a first pass later tonight.

File: vllm/model_executor/models/transformers.py (hunk 2 above)
Comment (ArthurZucker) [reply vllm/model_executor/models/transformers.py:298]: Arf, in my mind we would only need 1 "wrapper" to rule them all 😃 But makes sense. We can also have something like is multimodal, but might no work on your side!

**Review by ArthurZucker — COMMENTED** [review]
Have no say in this but that I am excited! 🚀

File: vllm/model_executor/models/transformers.py (hunk 1 above)
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:316]: yeah, making it a new model will self-solve the problem, so we can use 'sdpa' in `HFMultiModalModel`. We can work on refactoring later, it's been long on my list

File: vllm/model_executor/models/transformers.py [hunk 3]
```diff
@@ -252,11 +429,56 @@ def load_weights(self, weights: Iterable[tuple[str,
...
+            mask = (input_ids == self.config.image_token_index)
+            mask = mask.unsqueeze(-1).expand_as(inputs_embeds)
+            multimodal_embeddings = torch.cat(multimodal_embeddings)
+
+            # FIXME: The returned multimodal_embeddings must be either a 3D torch.Tensor of shape
+            # (num_items, feature_size, hidden_size), or a list / tuple of 2D torch.Tensor’s of shape
+            # (feature_size, hidden_size), so that multimodal_embeddings[i] retrieves the embeddings generated
+            # from the i-th multimodal data item (e.g, image) of the request.
+            inputs_embeds = inputs_embeds.masked_scatter(mask, multimodal_embeddings)
+        return inputs_embeds
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:484]: I am not familiar with Fuyu arch, so looking at Pixtral now I don't totally get why the outputs from `get_multimodal_embedding` contain image special tokens? Isn't it more straightforward to obtain only image related features and merge using mask (ids == special_image_token)?

Or maybe I am missing smth here

[the remark this answers is not in the document]
Comment (ywang96) [reply vllm/model_executor/models/transformers.py:484]: On V1 we require each embedding tensor corresponding to an image be contiguous and they must be **"sliced into"** the input embeddings instead of "scattered into".

The reason we're doing this is because V1 natively supports chunked prefilling, so when we schedule requests, we not only look at the decoder input sequence, but also when and how much of the encoder output/multimodal embeddings is needed which is why we need to track where they are in the input sequence. 

This location information is represented by [`PlaceholderRange`](https://github.com/vllm-project/vllm/blob/3f808cc0448bc60a4412575885db6ba8f5ab4671/vllm/multimodal/inputs.py#L111) generated by the input processor, which contains: 
(1) `offset`: the index where the multimodal embedding starts in the input sequence 
(2) `length`: how long it is (and obviously the actual size of output of `get_multimodal_embedding` for an image must match this).

Taking a toy example: my input sequence is `[T, T, T, T, I, I, I, I, I, T]` where `T` is text token and `I` is image placeholder token, then we have `offset=4, length=5`. Let's say in each step we schedule 3 tokens for prefilling:

Time 0: we schedule from `start_index=0` to `end_index=2`(`[T, T, T]`), since `2 < 4`, we know the `image_embeds` are not needed.

Time 1: we schedule to `start_index=3` to `end_index=5`(`[T, I, I]`), since `4 >= 4`, we know `image_embeds` is required for this step, so we run the vision encoder to generate it, and merge `image_embeds[:(5-4+1)]` into `[T, I, I]`), and we cache `image_embeds` becasue `end_index=5 + 1 = 6 < 4 + 5 = 9(offset + length)`, indicating it's still needed for the next chunked prefills.

Time 2: we schedule to `start_index=6` to `end_index=8`(`[I, I, I]`), since `8 >= 4`, we merge `image_embeds[(5-4+1):(8-4+1)]` into `[T, I, I]`), and now we can evict `image_embeds` since `end_index=8 + 1 = 9 >= 9`, indicating it's no longer needed for prefilling.

Time 3: we schedule the `start_index=9` to `end_index=9` (`[T]`), since `9 >= 9` we know `image_embeds` is not required for this step

As you can see this assumes the contiguous nature of multimodal embeddings. If we allow the embeddings to be scattered into the text embeddings, we would need a more complex definition for this location information, and thus also complicate the logic for scheduling.

Hope this helps!

File: vllm/model_executor/models/transformers.py (hunk 2 above)
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:298]: Is that already planned? Yeah, if that aligns with how VLLM works, it will help a lot to get the base model separately imo

Comment (hmellor) [reply vllm/model_executor/models/transformers.py:298]: It's not planned, no. But it wouldn't be hard to do in a separate PR.

I already have a PR which significantly changes the structure of `TransformersModel` (https://github.com/vllm-project/vllm/pull/12832 adds pipeline parallel support) which has been stuck because vLLM cannot use `transformers==4.49.0`. So I'd prefer that we get that merged before making big changes to `TransformersModel`

Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:298]: Ah I see. A side question: does that also mean we'll need two base models, one for TextOnly models and the other for Multimodal to register processors? And if yes as I understood from above comments, can we just inherit `MultimodalTransformersBase` from `TransformersBase` or are there any pitfalls for inheritance?

The reason I am asking, seems like the base model will grow bigger with more features, and maintaining two almost identical copies is harder than having only one version

File: vllm/model_executor/models/transformers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply vllm/model_executor/models/transformers.py:484]: Thanks, I will give it a look. The model code change in transformers is waiting review, so I will standardize processors to align with the way vLLM expects it

I'll assume the current format processors are handled won't change significantly in the near future

**Comment by zucchini-nlp** [comment]
The transformers repo was hugely refactored and now this PR is ready to be reviewed. I added a test and ran throughput benchmark on Vision-Arena with `disable_mm_preprocessor_cache=True` since transformers implementation doesn't support mm caching yet

```
Model: Llava-OV-0.5B
VLLM Throughput: 7.93 requests/s, 34115.47 total tokens/s, 1015.18 output tokens/s
HF Throughput: 7.50 requests/s, 32265.51 total tokens/s, 960.13 output tokens/s 
```

**Comment by DarkLight1337** [comment]
We need to upgrade to transformers 4.52 for this to work, correct?

**Comment by zucchini-nlp** [comment]
No, the patch won't work because I've been merging PRs after the release. We'll need to wait for the next release to make it work. Though I'd love to get it reviews in the meanwhile since I'm not sure how long the process might take

File: vllm/model_executor/model_loader/utils.py [hunk 4]
```diff
@@ -200,7 +200,7 @@ def resolve_transformers_arch(model_config: ModelConfig,
                 raise ValueError(
                     f"The Transformers implementation of {arch} is not "
                     "compatible with vLLM.")
-            architectures[i] = "TransformersForCausalLM"
+            architectures[i] = "TransformersForMultimodalLM"
```
Comment (DarkLight1337) [inline vllm/model_executor/model_loader/utils.py]: Hmm, won't this mean that the original `TransformersForCausalLM` isn't used anymore?

File: vllm/model_executor/models/transformers.py [hunk 5]
```diff
@@ -111,6 +120,260 @@ def replace_linear_class(
...
+    ) -> MultiModalInputs:
+        """
+        Process multi-modal inputs to be used in vLLM.
+
+        Apply HF Processor on prompt text and multi-modal data together,
+        outputting token IDs and processed tensors.
+        """
+        if return_mm_hashes:
+            raise ValueError(
+                "TransformersMultimodalLM doesn't support mm hashing yet! "
```
Comment (DarkLight1337) [inline vllm/model_executor/models/transformers.py]:
```suggestion
                "TransformersForMultimodalLM doesn't support mm hashing yet! "
```

File: vllm/v1/engine/mm_input_cache.py [hunk 6]
```diff
@@ -34,8 +34,8 @@ class MirroredProcessingCache:
 
     def __init__(self, model_config):
         mm_config = model_config.multimodal_config
-        disable_mm_preprocessor_cache = mm_config is not None and \
-            not mm_config.disable_mm_preprocessor_cache
+        disable_mm_preprocessor_cache = (
+            mm_config is not None and mm_config.disable_mm_preprocessor_cache)
```
Comment (DarkLight1337) [inline vllm/v1/engine/mm_input_cache.py]: Oh, I missed this bug, can you split it out to a bugfix PR?

File: vllm/model_executor/model_loader/utils.py (hunk 4 above)
Comment (Isotr0py) [reply vllm/model_executor/model_loader/utils.py]: I think we should only use `TransformersForMultimodalLM` for multimodal model. Perhaps we can assume multimodal model should have `vision_config`/`audio_config` in its hf_config?

File: tests/models/test_transformers.py [hunk 7]
```diff
@@ -71,6 +72,34 @@ def test_models(
...
 
+@pytest.mark.parametrize(
+    "model,model_impl",
+    [
+        ("llava-hf/llava-onevision-qwen2-0.5b-ov-hf",
+         "transformers"),  # dynamic image length and number of patches
+        ("HuggingFaceTB/SmolVLM-256M-Instruct",
+         "transformers"),  # has col/row special token between patches
+        ("Qwen/Qwen2.5-VL-3B-Instruct", "transformers"
+         ),  # pixel values from processor are not 4D or 5D arraya
```
Comment (Isotr0py) [inline tests/models/test_transformers.py]: Hmmm, seems this test doesn't cover `auto` model_impl fallback... Perhaps we can add a `transformers` multimodal model hasn't supported in vLLM (like `stepfun-ai/GOT-OCR-2.0-hf`) to cover the fallback?

File: vllm/model_executor/model_loader/utils.py (hunk 4 above)
Comment (zucchini-nlp) [reply vllm/model_executor/model_loader/utils.py]: Yeah, this is one issue I am discussing with @hmellor internally. I couldn't make it load the correct model, unless hardcoded to fallback for `TransformersForMultimodalLM`

We can try to infer if model supports multimodal by a few assumptions, though it won't be 100% robust. Yeah, like checking the text/vision-config or if there exists a processor for model

File: tests/models/test_transformers.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/models/test_transformers.py]: oke, sounds good

File: vllm/model_executor/model_loader/utils.py (hunk 4 above)
Comment (zucchini-nlp) [reply vllm/model_executor/model_loader/utils.py]: added a verification step based on config for now

File: vllm/model_executor/models/transformers.py [hunk 8]
```diff
@@ -506,3 +746,172 @@ def load_weights(self, weights: Iterable[tuple[str,
...
         )
         return loader.load_weights(weights, mapper=self.hf_to_vllm_mapper)
+
+
+@MULTIMODAL_REGISTRY.register_processor(
+    MultiModalProcessor,
+    info=MultiModalProcessingInfo,
+    dummy_inputs=MultiModalDummyInputsBuilder)
+class TransformersForMultimodalLM(nn.Module, SupportsQuant, SupportsLoRA,
+                                  SupportsPP, SupportsMultiModal):
```
Comment (Isotr0py) [inline vllm/model_executor/models/transformers.py:757]: For VLM, we need to implement `get_mm_mapping` as well to support LoRA, right? @jeejeelee

**Comment by zucchini-nlp** [comment]
Nooo, I hate rebasing issues 😢

**Comment by zucchini-nlp** [comment]
Closing in favor of https://github.com/vllm-project/vllm/pull/20543
